MSTR-37 : [Feat/37] 드롭다운 고도화

Kim-Hyunjo opened the pull request:
![ezgif com-gif-maker (1)](https://user-images.githubusercontent.com/63906230/177110906-5a2ce34d-d9ed-4474-9e8c-5bdc2f78105c.gif)

- 외부 영역 클릭시 숨기기
- 한번에 드롭다운 한개만 선택되게 하기
- 각 드롭다운 요소들 체크 여부 관리하기

Kim-Hyunjo committed:
✨ Add listenOutsideClick function

- menuRef가 가르키는 dom요소(드롭다운)가 아닌 곳이 클릭되면 isOpen false로 바꾸기

Kim-Hyunjo committed:
🏷  Add Dropdown Element type

- 태그 선택시 사용되는 드롭다운 요소의 type을 정의함
- id, name 가짐

Kim-Hyunjo committed:
🏷 Add Taglist, Tagtypelist

- 태그 타입 리스트는 태그 종류를 정의함
- 태그 리스트는 각각의 태그 종류에 따른 내용을 담고 있음

Kim-Hyunjo committed:
♻️ Change Dropdown file loc && seperate element

- 드롭다운 컴포넌트 파일의 위치를 옮김과 동시에 내부 요소 하나하나를 분리
- 체크박스 각각의 체크 여부 상위 드롭다운 컴포넌트에서 관리
- 추후 드롭다운 리스트에서 복수 조건에 대한 체크여부 관리 예정

Kim-Hyunjo committed:
♻️ Refactor Dropdown List

- 무지성 string으로 되어있던 드롭다운들을 TAGTYPELIST를 이용하여 데이터 prop으로 전달과 동시에 맵핑

Kim-Hyunjo committed:
🔥 Remove Unused Tag

Kim-Hyunjo committed:
💄 Add media query for screen under width 600px

- html font size 6px로 설정
- 추후 구체적인 반응형 설정 예정

Kim-Hyunjo committed:
🚸 Change px -> rem in Problem List Page

kshired commented on src/Component/Utils/Dropdown/index.tsx:
```diff
@@ -0,0 +1,110 @@
+              id={e.id}
+              name={e.name}
+              checkedItemHandler={checkedItemHandler}
+              key={e.id + e.name}
```
`e.id + e.name` 대신 `${e.id}${e.name}`을 써도 좋을 것 같아요

kshired replied:
그리고 key를 넣어서 li 요소들의 최적화를 하는 것 같은데, 이 key도 의미를 분명하게 하는게 도움이 될까요? 
( 잘 모름, 진짜 물어보는 것 )

Kim-Hyunjo replied:
각 드롭다운 요소마다 가진 아이디값이 고유할거라고 예상하지만 리액트에서 list의 key값은 렌더링시의 식별자라서 혹시라도 겹치면 리렌더링시에 요소가 사라지는 위험이 발생할 수도 있기 때문에 확신을 위해서 id와 name을 합쳐서 사용했습니다.

참고로 key값이 길어지는 것이 성능에 영향을 주지는 않는다고 하네요! (🔗 [관련 링크](https://stackoverflow.com/questions/41380976/react-performance-implications-of-long-key-value-on-component))

kshired commented on src/Component/Utils/Dropdown/index.tsx:
```diff
@@ -0,0 +1,110 @@
+    if (isChecked) {
+      checkedItems.add(id);
+      setCheckedItems(checkedItems);
+    } else if (!isChecked && checkedItems.has(id)) {
```
id가 없는 아이템이 체크될 일이 있나요?

kshired replied:
생각해보니 일단 checkedItems set 로직부터 수정이 필요할듯합니다~

useState 사용시에 immutability가 중요한 걸로 아는데, 아래로 변경하는게 좋지 않을까요?

## Add 의 경우
```js
const newCheckedItems = [...checkedItems, id]
setCheckedItems(new Set(newCheckedItems))
```

## Delete의 경우
```js
const newCheckedItems = [...checkedItems].filter( item => item !== id )
setCheckedItems(new Set(newCheckedItems))
```

## 참고
Set 자료구조에서 delete는 존재하지 않는 원소에 대해서, delete 시도시 오류를 throw하지 않고 boolean으로  삭제 성공 여부를 return한다고 하니 굳이 `checkedItems.has(id)`를 할 필요는 없을 듯 합니다.

어차피 filter 처리하면 빠져도 되는 로직일거구요~

kshired commented on src/Component/Utils/Dropdown/DropdownElement.tsx:
```diff
@@ -0,0 +1,38 @@
+
+  return (
+    <li css={dropdownContentElementStyle}>
+      <input type="checkbox" name={name} id={id} onChange={(e) => checkHandler(e)} />
```
`onChange={checkHandler}` 처럼 할 수 있지 않을까요

Kim-Hyunjo replied:
세상에 그렇네요 감사합니다 선생님!!!!!

kshired commented on src/Component/Utils/listenOutsideClick.ts:
```diff
@@ -0,0 +1,19 @@
+  return () => {
+    if (!menuRef.current) return;
+
+    ['click', 'touchstart'].forEach((e) => {
```
얘네도 constant로 빼놓는게 나중에 refactoring할 때 편할수도

Kim-Hyunjo replied:
오 굿아이디어네요!! 반영했습니다 감사합니다!!

kshired reviewed (commented):
참고해주세요~

Kim-Hyunjo committed:
✏️ Fix string concatenate typo

- plus -> template literal

Kim-Hyunjo committed:
✏️ Fix function typo

- remove unnecessary code

Kim-Hyunjo committed:
🤖 Add EVENT constant

- click, touchstart event 명 상수 추가
- tag관련 상수 분리

Kim-Hyunjo committed:
🐛 Fix state immutability issue

- useState 사용시 set을 직접 수정할 경우 immutability 망가지는 문제 수정
- set 자료구조에서 delete는 오류를 발생시키지 않으므로 id 존재 여부 체크 필요 없음

kshired commented on src/Component/Utils/listenOutsideClick.ts:
```diff
@@ -7,7 +8,7 @@ export default function listenOutsideClick( //menuRef가 가르키는 dom요소(
     if (!menuRef.current) return;
 
-    ['click', 'touchstart'].forEach((e) => {
+    [EVENT.CLICK, EVENT.TOUCHSTART].forEach((e) => {
```
EVENT는 나중에 추가 될 일이 있어서 CLICK, TOUCHSTART만 넣은걸까요?

kshired reviewed (approved):
LGTM!
